### PR TITLE
Split superpage allocation

### LINKED ISSUE #85: Split num_blocks when allocating superpages
In [nova_alloc_superpage](https://github.com/NVSL/linux-nova/blob/master/fs/nova/balloc.c#L589),  it's trying to find a free_list node that can hold all **consecutive** num_blocks. It's hard to succeed when fallocating a large space. For example, even if there is two 800MB free_list nodes, it will fail to use superpages when fallocating 1.6GB space. 
I think free blocks for superpages can be obtained from different free_list nodes. It can be done by directly returning the number of 2M-aligned free blocks found in nova_alloc_superpage, just [like when allocating small pages](https://github.com/NVSL/linux-nova/blob/master/fs/nova/balloc.c#L835).

### PR DESCRIPTION
**Pull request by pleiadesian**
fixed #85

### CODE REVIEW

**Comment by pleiadesian** [comment]
I've tested with a simple fallocate program which fallocates 2G and accesses pgoff 512 and 513. According to [trace log](https://gist.github.com/pleiadesian/d6d16e33a552b6d351807b07d7e8140e#file-trace-log), super pages are allocated and page fault is only triggered at pgoff 512.

**Comment by Andiry** [comment]
Thanks, I will take a look.

Andiry

File: fs/nova/balloc.c [hunk 1]
```diff
@@ -746,8 +752,12 @@ static long nova_alloc_blocks_in_free_list(struct super_block *sb,
...
 	if (IS_DATABLOCKS_2MB_ALIGNED(num_blocks, atype)) {
-		found_hugeblock = nova_alloc_superpage(sb, free_list,
+		ret_blocks = nova_alloc_superpage(sb, free_list,
 					num_blocks, new_blocknr, from_tail);
+		if (ret_blocks > 0 && new_blocknr != 0) {
+			num_blocks = ret_blocks;
+			found_hugeblock = 1;
+		}
 		if (found_hugeblock)
 			goto success;
```
Comment (Andiry) [inline fs/nova/balloc.c:762]: Remove this found_hugeblock check and move goto success to the previous if block.

File: fs/nova/balloc.c [hunk 2]
```diff
@@ -746,8 +752,12 @@ static long nova_alloc_blocks_in_free_list(struct super_block *sb,
 
 	/* Try huge block allocation for data blocks first */
 	if (IS_DATABLOCKS_2MB_ALIGNED(num_blocks, atype)) {
-		found_hugeblock = nova_alloc_superpage(sb, free_list,
+		ret_blocks = nova_alloc_superpage(sb, free_list,
 					num_blocks, new_blocknr, from_tail);
+		if (ret_blocks > 0 && new_blocknr != 0) {
```
Comment (Andiry) [inline fs/nova/balloc.c:757]: new_blocknr is a pointer. You want to check *new_blocknr?

File: fs/nova/balloc.c [hunk 3]
```diff
@@ -586,7 +586,7 @@ struct nova_range_node *nova_alloc_blocknode_atomic(struct super_block *sb)
 #define IS_DATABLOCKS_2MB_ALIGNED(numblocks, atype) \
 		(!(num_blocks & PAGES_PER_2MB_MASK) && (atype == DATA))
 
-bool nova_alloc_superpage(struct super_block *sb,
+static long nova_alloc_superpage(struct super_block *sb,
```
Comment (Andiry) [inline fs/nova/balloc.c:589]: Add a comment that this method returns the number of blocks allocated.

File: fs/nova/balloc.c [hunk 4]
```diff
@@ -627,6 +627,11 @@ bool nova_alloc_superpage(struct super_block *sb,
 		 *    requested blocks.
 		 * Otherwise, we are free to go.
 		 */
+		if (num_blocks > (curr_blocks - left_margin)) {
```
Comment (Andiry) [inline fs/nova/balloc.c:630]: Add a comment that we are doing best effort here to allocate as many 2MB blocks as possible.

**Review by Andiry — CHANGES_REQUESTED** [review]

File: fs/nova/balloc.c (hunk 2 above)
Comment (pleiadesian) [reply fs/nova/balloc.c:757]: Yes, I'll change it to *new_blocknr != 0.

**Comment by pleiadesian** [comment]
Merged 3 commits into one single commit.

**Comment by Andiry** [comment]
Thanks!

**Pull request merged**